tests/integration/run.py

```diff
@@ -0,0 +1,984 @@
+    include_tags: list[str] | None = None,
+    exclude_tags: list[str] | None = None,
+) -> list[Case] | None:
+    """Load the test cases from the test case directories.
+
+    Parameters
+    ----------
+    test_case_dirs : list[str]
+        Test case directores.
+    check_expected_result_files : bool
+        Whether to check if expected result files are valid.
+    include_tags : list[str] | None
+        A selected test case must contain all of these tags.
+    exclude_tags : list[str] | None
+        A selected test case must not contain any of these tags.
+    """
+    include_tags = include_tags or []
+    exclude_tags = exclude_tags or []
+
+    err = False
+    test_cases: list[Case] = []
+
+    for test_case_dir in test_case_dirs:
+        try:
+            case_config = load_config(test_case_dir, check_expected_result_files)
+        except InvalidConfigError as exc:
+            logger.error("Case '%s' fails validation.", test_case_dir)
+            logger.error(exc.error_msg)
+            err = True
+        else:
+            # Each --include-tag/--exclude-tag argument adds an additional constraint
+            # that a selected test case needs to satisfy, i.e. a selected test case must:
+            # - contains all tags specified with --include-tag
+            # - contains no tag specified with --exclude-tag
+            select_case = True
+            for include_tag in include_tags:
+                if include_tag not in case_config["tags"]:
+                    logger.info(
+                        "Skipping case '%s' for not having tag '%s'.",
+                        *(test_case_dir, include_tag),
+                    )
+                    select_case = False
+                    break
+            for exclude_tag in exclude_tags:
+                if exclude_tag in case_config["tags"]:
+                    logger.info(
+                        "Skipping case '%s' for having tag '%s'",
+                        *(test_case_dir, exclude_tag),
+                    )
+                    select_case = False
+                    break
+            if not select_case:
+                continue
+            steps = []
+            for step_id, step_config in enumerate(case_config["steps"]):
+                step = parse_step_config(step_id, step_config)
+                steps.append(step)
+            test_case = Case(
+                case_dir=test_case_dir,
+                description=case_config["description"],
+                steps=steps,
+            )
+            test_cases.append(test_case)
+            logger.info("Case '%s' passes validation.", test_case_dir)
+
+    if err:
+        return None
+
+    return test_cases
+
+
+def do_check(test_case_dirs: list[str], check_expected_result_files: bool) -> int:
+    """Execute the check command."""
+    test_cases = load_test_cases(
+        test_case_dirs,
+        check_expected_result_files=check_expected_result_files,
+    )
+    if test_cases is None:
```

Comment:
nit
```suggestion
    if test_cases is None:
        logger.error("Error encountered while loading test config.")
```

Reply:
In 63829f3c6481f007d68a91632f92b3fe02fc4143, I have moved error-logging into the `load_test_cases` function. Consumers of this function should not need to add additional logging.